src/interface/trsm.hpp

```diff
@@ -0,0 +1,432 @@
+  const index_t localMemSize = blockSize * blockSize;
+
+  // Instantiate the appropriate diagonal blocks inversion based on the matrix
+  // type
+  typename executor_t::policy_t::event_t invertBlocksEvent;
+  if (isUnitDiag && isUpper) {
+    auto diagInverter =
+        make_diag_blocks_inverter<true, true>(bufferA, bufferInvA);
+    invertBlocksEvent =
+        ex.execute(diagInverter, localSize, globalSize, localMemSize);
+  } else if (!isUnitDiag && isUpper) {
+    auto diagInverter =
+        make_diag_blocks_inverter<false, true>(bufferA, bufferInvA);
+    invertBlocksEvent =
+        ex.execute(diagInverter, localSize, globalSize, localMemSize);
+  } else if (isUnitDiag && !isUpper) {
+    auto diagInverter =
+        make_diag_blocks_inverter<true, false>(bufferA, bufferInvA);
+    invertBlocksEvent =
+        ex.execute(diagInverter, localSize, globalSize, localMemSize);
+  } else if (!isUnitDiag && !isUpper) {
+    auto diagInverter =
+        make_diag_blocks_inverter<false, false>(bufferA, bufferInvA);
+    invertBlocksEvent =
+        ex.execute(diagInverter, localSize, globalSize, localMemSize);
+  }
+
+  // Creates a copy of B to avoid overwriting the input in GEMM. While computing
+  // output X will hold the TRSM result and will be copied to B at the end
+  const index_t BSize = ldb * (N - 1) + M;
+  const index_t ldx = ldb;
+  auto X = make_sycl_iterator_buffer<element_t>(BSize);
+  auto bufferX = make_matrix_view<col_major>(ex, X, M, N, ldx);
+  trsmEvents = concatenate_vectors(
+      trsmEvents, ex.execute(make_op<Assign>(bufferX, bufferB), BSize));
+
+  if (isLeft) {
+    if ((isUpper && isTranspose) || (!isUpper && !isTranspose)) {
+      // Solves the system AX = alpha*B, as described in the documentation of
+      // the function when X is lower triangular
+
+      // True when (lower triangular) or (upper triangular and transposed)
+      for (index_t i = 0; i < M; i += blockSize) {
+        const index_t currentBlockSize = std::min(M - i, blockSize);
+        index_t gemmM = currentBlockSize;
+        index_t gemmN = N;
+        index_t gemmK = currentBlockSize;
+        char gemmTransA = isTranspose ? 't' : 'n';
+        char gemmTransB = 'n';
+        element_t gemmAlpha = (i == 0) ? alpha : element_t{1};
+        element_t gemmBeta = element_t{0};
+        index_t gemmLda = blockSize;
+        index_t gemmLdb = ldb;
+        index_t gemmLdc = ldx;
+        std::ptrdiff_t offsetA = i * blockSize;
+        std::ptrdiff_t offsetB = i;
+        std::ptrdiff_t offsetC = i;
+        auto gemmBufferA = invA + offsetA;
+        auto gemmBufferB = B + offsetB;
+        auto gemmBufferC = X + offsetC;
+        auto gemmEvent =
+            internal::_gemm(ex, gemmTransA, gemmTransB, gemmM, gemmN, gemmK,
+                            gemmAlpha, gemmBufferA, gemmLda, gemmBufferB,
+                            gemmLdb, gemmBeta, gemmBufferC, gemmLdc);
+        trsmEvents = concatenate_vectors(trsmEvents, gemmEvent);
```

Comment:
These operations are repeated multiple times. Can you wrap them in a function and call the function instead? you can create a function `call_gemm` to do the operation

Reply:
What would be the advantage of wrapping this in a function? All this is doing is calling gemm with different parameters depending on which block should be multiplied. Having this is a function would just encapsulate gemm.

Reply:
I was thinking the same thing when I first saw the code, but I realized most of the code is setting up parameters so it doesn't seem like the amount of code can be reduced.

Reply:
What @mehdi-goli wants is for all the parameters to be inline in the `_gemm` call. I don't like this.

Reply:
There are lots of boilerplate variable here which is not needed. e.g `int gemmN=N;` is boilerplate. `int gemmLdb=ldb;` is boilerplate, etc. there are full of them. if you write a call to gemm function`_gemm(...., currentBlock, N, currenrtBlock,..., B+i, ....);` everyone who is familiar with linear algebra and gemm understand what it says. GEMM algorithm and the call to it is standard. This is part of the internal code and anyone works with it is expected to understand gemm and linear algebra. if you want to be expressive you should put some comments on top of each block and explain which block each gemm is computing.

Reply:
> There are lots of boilerplate variable here which is not needed

This boilerplate code, despite not being strictly needed makes the code a lot more readable than a call to a function that has 16 inline parameters. It was a lot easier to write this code having all the parameters in separate variables.

> everyone who is familiar with linear algebra and gemm understand what it says

This is an assumption that breaks really easily. Yes, there is a call to gemm there but where it is operating and what is it doing is far from obvious. It is a lot easier to see what the value of each argument is, which in turns makes it easier to think about what the operation is doing

> GEMM algorithm and the call to it is standard

Being standard doesn't mean it should be difficult to read.

> This is part of the internal code and anyone works with it is expected to understand gemm and linear algebra

Exactly because it is internal code it should be easier to maintain. 

> you should put some comments on top of each block and explain which block each gemm is computing

I agree with this, good comments are always good, but readable code requires less comments because the code is easier to read and understand.

With this said, I will proceed and make the change you want but I still don't like it.

Reply:
changes are in 108700435ecdf3742e87b6e9c869c03be20e0881

Reply:
IMO it's more readable this way (with named arguments) because you don't have to go to the definition of the GEMM function to figure out what each one does.

Reply:
> With this said, I will proceed and make the change you want but I still don't like it.
Thank you

Reply:
> I agree with this, good comments are always good,
Let me know when the comment is added so we can resolve this issue

Reply:
> IMO it's more readable this way (with named arguments) because you don't have to go to the definition of the GEMM function to figure out what each one does.

If it is too hard to read gemm parameters, we can comment out gemm definition on top of the code rather than adding boilerplates. In this case, you don't have boilerplates and can easily see one to one mapping of the parameters without going to the gemm definition

Reply:
Comments added in cfcd84278807ab512c361f6bfe60f2eedccef95f